base/floatfuncs.jl

```diff
@@ -20,6 +20,7 @@ maxintfloat(::Type{Float64}) = 9007199254740992.
 maxintfloat(::Type{Float32}) = Float32(16777216.)
 maxintfloat(::Type{Float16}) = Float16(2048f0)
 maxintfloat(x::T) where {T<:AbstractFloat} = maxintfloat(T)
+maxintfloat(::Type{S}, ::Type{T}) where {S<:AbstractFloat, T<:Integer} = (m = maxintfloat(S); Int === Int32 ? min(m, S(typemax(T))) : m)
```

Comment:
Thanks Tim. I had considered adding the extra generality, but wasn't sure whether it would slow down the code. Turns out I should have trusted LLVM's cleverness even more. Thanks for these suggestions, very informative!